File: starlore-go/internal/services/article_service.go
package services

import (
	"encoding/json"
	"math"
	"net/http"
	"time"

	"starlore-go/internal/database"
	"starlore-go/internal/models"
	"starlore-go/internal/schemas"
	"starlore-go/internal/utils"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"
)

// ListArticles 文章列表
func ListArticles(c *gin.Context) {
	page := utils.ParseInt(c.Query("page"), 1)
	limit := utils.ParseInt(c.Query("limit"), 10)
	category := c.Query("category")
	search := c.Query("search")
	tag := c.Query("tag")
	uid, _ := c.Get("userId")

	query := database.DB.Model(&models.Article{})

	if uid != nil {
		query = query.Where("user_id = ? AND status = ?", uid, "published")
	}

	if category != "" && category != "全部" {
		query = query.Where("category = ?", category)
	}
	if search != "" {
		query = query.Where("title LIKE ? OR description LIKE ?", "%"+search+"%", "%"+search+"%")
	}
	if tag != "" {
		query = query.Where("JSON_CONTAINS(tags, JSON_ARRAY(?))", tag)
	}

	var total int64
	query.Count(&total)
	pages := int(math.Ceil(float64(total) / float64(limit)))

	var articles []models.Article
	query.Order("createdAt DESC").Offset((page - 1) * limit).Limit(limit).Find(&articles)

	// 批量获取作者名（避免 N+1）
	userIDs := make(map[int]bool)
	for _, a := range articles {
		userIDs[a.UserID] = true
	}
	authorMap := make(map[int]string)
	if len(userIDs) > 0 {
		ids := make([]int, 0, len(userIDs))
		for id := range userIDs {
			ids = append(ids, id)
		}
		var users []models.User
		database.DB.Where("id IN ?", ids).Find(&users)
		for _, u := range users {
			name := u.Nickname.String
			if name == "" {
				name = u.Username
			}
			authorMap[u.ID] = name
		}
	}

	summaries := make([]schemas.ArticleSummary, len(articles))
	for i, a := range articles {
		summaries[i] = toArticleSummary(a, authorMap[a.UserID])
	}

	c.JSON(http.StatusOK, schemas.ArticleListResponse{
		Data: summaries,
		Pagination: schemas.PaginationInfo{
			Current: page,
			Total:   total,
			Pages:   pages,
		},
	})
}

// GetArticle 获取文章详情
func GetArticle(c *gin.Context) {
	id := utils.ParseInt(c.Param("id"), 0)
	uid, _ := c.Get("userId")

	var article models.Article
	query := database.DB.Where("id = ?", id)
	if uid != nil {
		query = query.Where("user_id = ?", uid)
	}
	if err := query.First(&article).Error; err != nil {
		utils.Fail(c, http.StatusNotFound, "文章不存在")
		return
	}

	// 使用 SQL 表达式避免竞态条件
	database.DB.Model(&article).UpdateColumn("view_count", gorm.Expr("view_count + 1"))
	article.ViewCount++

	c.JSON(http.StatusOK, toArticleDetail(article))
}

// CreateArticle 创建文章
func CreateArticle(c *gin.Context, req schemas.CreateArticleRequest) {
	uid := utils.GetUserID(c)
	if uid == 0 {
		utils.Fail(c, http.StatusUnauthorized, "未登录")
		return
	}

	now := time.Now()
	tagsJSON := "[]"
	if req.Tags != nil {
		b, err := json.Marshal(req.Tags)
		if err == nil {
			tagsJSON = string(b)
		}
	}

	article := models.Article{
		UserID:      uid,
		Title:       req.Title,
		Content:     req.Content,
		Description: req.Description,
		Category:    req.Category,
		Tags:        tagsJSON,
		CoverImage:  req.CoverImage,
		ViewCount:   0,
		Status:      "published",
		CreatedAt:   &now,
		UpdatedAt:   &now,
	}
	if article.Category == "" {
		article.Category = "随笔"
	}
	if req.Status != "" {
		article.Status = req.Status
	}

	database.DB.Create(&article)
	updateCategoryCount(article.Category, uid)

	utils.Ok(c, "文章创建成功", gin.H{"id": article.ID})
}

// UpdateArticle 更新文章
func UpdateArticle(c *gin.Context, req schemas.UpdateArticleRequest) {
	uid := utils.GetUserID(c)
	id := utils.ParseInt(c.Param("id"), 0)

	var article models.Article
	if err := database.DB.Where("id = ? AND user_id = ?", id, uid).First(&article).Error; err != nil {
		utils.Fail(c, http.StatusNotFound, "文章不存在")
		return
	}

	oldCategory := article.Category
	updates := map[string]interface{}{}
	if req.Title != nil {
		updates["title"] = *req.Title
	}
	if req.Content != nil {
		updates["content"] = *req.Content
	}
	if req.Description != nil {
		updates["description"] = *req.Description
	}
	if req.Category != nil {
		updates["category"] = *req.Category
	}
	if req.Tags != nil {
		tagsJSON, _ := json.Marshal(req.Tags)
		updates["tags"] = string(tagsJSON)
	}
	if req.CoverImage != nil {
		updates["cover_image"] = *req.CoverImage
	}
	if req.Status != nil {
		updates["status"] = *req.Status
	}
	now := time.Now()
	updates["updatedAt"] = &now

	database.DB.Model(&article).Updates(updates)

	if req.Category != nil && *req.Category != oldCategory {
		updateCategoryCount(oldCategory, article.UserID)
		updateCategoryCount(*req.Category, article.UserID)
	}

	utils.Ok(c, "文章更新成功", nil)
}

// DeleteArticle 删除文章
func DeleteArticle(c *gin.Context) {
	uid := utils.GetUserID(c)
	id := utils.ParseInt(c.Param("id"), 0)

	var article models.Article
	if err := database.DB.Where("id = ? AND user_id = ?", id, uid).First(&article).Error; err != nil {
		utils.Fail(c, http.StatusNotFound, "文章不存在")
		return
	}

	database.DB.Delete(&article)
	updateCategoryCount(article.Category, article.UserID)

	utils.Ok(c, "文章删除成功", nil)
}

// GetBlogStats 获取博客统计
func GetBlogStats(c *gin.Context) {
	uid := utils.GetUserID(c)

	var totalArticles int64
	database.DB.Model(&models.Article{}).Where("user_id = ? AND status = ?", uid, "published").Count(&totalArticles)

	var totalCategories int64
	database.DB.Model(&models.Category{}).Where("user_id = ?", uid).Count(&totalCategories)

	var totalViews int64
	database.DB.Model(&models.Article{}).Where("user_id = ? AND status = ?", uid, "published").
		Select("COALESCE(SUM(view_count), 0)").Scan(&totalViews)

	var popularArticles []models.Article
	database.DB.Where("user_id = ? AND status = ?", uid, "published").
		Order("createdAt DESC").Limit(4).Find(&popularArticles)

	articleSummaries := make([]schemas.ArticleSummary, len(popularArticles))
	for i, a := range popularArticles {
		articleSummaries[i] = toArticleSummary(a, "")
	}

	var popularCategories []schemas.CategoryInfo
	database.DB.Model(&models.Category{}).Where("user_id = ?", uid).
		Order("article_count DESC").Limit(5).
		Select("id, name, article_count").Scan(&popularCategories)

	c.JSON(http.StatusOK, gin.H{
		"data": schemas.BlogStatsData{
			TotalArticles:     totalArticles,
			TotalCategories:   totalCategories,
			TotalViews:        totalViews,
			PopularArticles:   articleSummaries,
			PopularCategories: popularCategories,
		},
	})
}

// GetDailyStats 获取每日统计
func GetDailyStats(c *gin.Context) {
	uid := utils.GetUserID(c)

	type dailyRow struct {
		Date  string
		Count int
	}
	var rows []dailyRow

	startDate := time.Now().AddDate(0, 0, -6).Format("2006-01-02")
	database.DB.Model(&models.Article{}).
		Where("user_id = ? AND DATE(createdAt) >= ?", uid, startDate).
		Select("DATE(createdAt) as date, COUNT(*) as count").
		Group("DATE(createdAt)").Scan(&rows)

	dateMap := make(map[string]int)
	for _, r := range rows {
		dateMap[r.Date] = r.Count
	}

	data := make([]schemas.DailyItem, 7)
	for i := 0; i < 7; i++ {
		d := time.Now().AddDate(0, 0, -6+i)
		ds := d.Format("2006-01-02")
		data[i] = schemas.DailyItem{Date: ds, Count: dateMap[ds]}
	}

	c.JSON(http.StatusOK, gin.H{"data": data})
}

func updateCategoryCount(categoryName string, userID int) {
	if categoryName == "" {
		return
	}
	var count int64
	database.DB.Model(&models.Article{}).
		Where("category = ? AND user_id = ? AND status = ?", categoryName, userID, "published").
		Count(&count)

	var category models.Category
	if database.DB.Where("name = ? AND user_id = ?", categoryName, userID).First(&category).Error == nil {
		database.DB.Model(&category).Update("article_count", count)
	} else {
		now := time.Now()
		database.DB.Create(&models.Category{
			UserID:       userID,
			Name:         categoryName,
			ArticleCount: int(count),
			CreatedAt:    &now,
			UpdatedAt:    &now,
		})
	}
}

func toArticleSummary(a models.Article, authorName string) schemas.ArticleSummary {
	return schemas.ArticleSummary{
		ID:          a.ID,
		UserID:      a.UserID,
		AuthorName:  authorName,
		Title:       a.Title,
		Status:      a.Status,
		Description: a.Description,
		Category:    a.Category,
		Tags:        utils.ParseTags(a.Tags),
		CoverImage:  a.CoverImage,
		ViewCount:   a.ViewCount,
		CreatedAt:   a.CreatedAt,
	}
}

func toArticleDetail(a models.Article) schemas.ArticleDetail {
	return schemas.ArticleDetail{
		ID:          a.ID,
		Title:       a.Title,
		Content:     a.Content,
		Description: a.Description,
		Category:    a.Category,
		Tags:        utils.ParseTags(a.Tags),
		CoverImage:  a.CoverImage,
		ViewCount:   a.ViewCount,
		Status:      a.Status,
		CreatedAt:   a.CreatedAt,
		UpdatedAt:   a.UpdatedAt,
	}
}

File: starlore-py/app/services/article_service.py
"""文章服务：CRUD、分页、搜索、统计。"""

import logging
import math
from datetime import datetime, timedelta

from sqlalchemy import select, func, or_, text
from sqlalchemy.ext.asyncio import AsyncSession

from app.exceptions import BadRequestException, NotFoundException
from app.models.article import Article
from app.models.category import Category
from app.models.user import User
from app.schemas.article import (
    ArticleDetail,
    ArticleListResponse,
    ArticleSummary,
    CreateArticleRequest,
    UpdateArticleRequest,
)
from app.schemas.common import PaginationInfo

logger = logging.getLogger(__name__)


def _to_summary(article: Article, author_name: str | None = None) -> ArticleSummary:
    return ArticleSummary(
        id=article.id,
        userId=article.user_id,
        authorName=author_name,
        title=article.title,
        status=article.status,
        description=article.description,
        category=article.category,
        tags=article.tags,
        cover_image=article.cover_image,
        view_count=article.view_count,
        is_public=article.is_public,
        createdAt=article.createdAt,
    )


def _to_detail(article: Article) -> ArticleDetail:
    return ArticleDetail(
        id=article.id,
        title=article.title,
        content=article.content,
        description=article.description,
        category=article.category,
        tags=article.tags,
        cover_image=article.cover_image,
        view_count=article.view_count,
        status=article.status,
        is_public=article.is_public,
        createdAt=article.createdAt,
        updatedAt=article.updatedAt,
    )


async def get_all_articles(
    db: AsyncSession,
    user_id: int | None = None,
    page: int = 1,
    limit: int = 10,
    category: str | None = None,
    search: str | None = None,
    tag: str | None = None,
    filter_user_id: int | None = None,
    published_only: bool = True,
) -> ArticleListResponse:
    """获取文章列表（分页、搜索、筛选）。"""
    query = select(Article)

    # 用户过滤
    if filter_user_id is not None:
        query = query.where(Article.user_id == filter_user_id)
    elif not published_only:
        pass  # admin 模式不过滤
    elif user_id is not None:
        query = query.where(Article.user_id == user_id, Article.status == "published")
    else:
        query = query.where(Article.status == "published")

    # 分类过滤
    if category and category != "全部":
        query = query.where(Article.category == category)

    # 搜索过滤
    if search:
        query = query.where(
            or_(
                Article.title.ilike(f"%{search}%"),
                Article.description.ilike(f"%{search}%"),
            )
        )

    # 标签过滤（使用参数化查询防止 SQL 注入）
    if tag:
        from sqlalchemy import literal
        query = query.where(text("JSON_CONTAINS(tags, JSON_ARRAY(:tag))").bindparams(tag=tag))

    # 计算总数
    count_query = select(func.count()).select_from(query.subquery())
    total_result = await db.execute(count_query)
    total = total_result.scalar() or 0
    pages = math.ceil(total / limit) if limit > 0 else 0

    # 分页查询
    offset = (page - 1) * limit
    query = query.order_by(Article.createdAt.desc()).offset(offset).limit(limit)
    result = await db.execute(query)
    articles = list(result.scalars().all())

    # 批量获取作者名
    user_ids = list({a.user_id for a in articles})
    author_map: dict[int, str] = {}
    if user_ids:
        users_result = await db.execute(select(User).where(User.id.in_(user_ids)))
        for u in users_result.scalars().all():
            author_map[u.id] = u.nickname or u.username

    summaries = [_to_summary(a, author_map.get(a.user_id)) for a in articles]
    return ArticleListResponse(
        data=summaries,
        pagination=PaginationInfo(current=page, total=total, pages=pages),
    )


async def get_public_articles(
    db: AsyncSession,
    page: int = 1,
    limit: int = 10,
    category: str | None = None,
    search: str | None = None,
    tag: str | None = None,
) -> ArticleListResponse:
    """List only published articles explicitly marked as public."""
    query = select(Article).where(Article.status == "published", Article.is_public.is_(True))
    if category and category != "全部":
        query = query.where(Article.category == category)
    if search:
        query = query.where(or_(Article.title.ilike(f"%{search}%"), Article.description.ilike(f"%{search}%")))
    if tag:
        query = query.where(text("JSON_CONTAINS(tags, JSON_ARRAY(:tag))").bindparams(tag=tag))

    total = (await db.execute(select(func.count()).select_from(query.subquery()))).scalar() or 0
    pages = math.ceil(total / limit) if limit else 0
    result = await db.execute(
        query.order_by(Article.createdAt.desc()).offset((page - 1) * limit).limit(limit)
    )
    articles = list(result.scalars().all())
    user_ids = {article.user_id for article in articles}
    author_map: dict[int, str] = {}
    if user_ids:
        users = await db.execute(select(User).where(User.id.in_(user_ids)))
        author_map = {user.id: user.nickname or user.username for user in users.scalars().all()}
    return ArticleListResponse(
        data=[_to_summary(article, author_map.get(article.user_id)) for article in articles],
        pagination=PaginationInfo(current=page, total=total, pages=pages),
    )


async def get_public_article_by_id(db: AsyncSession, article_id: int) -> ArticleDetail:
    result = await db.execute(
        select(Article).where(
            Article.id == article_id,
            Article.status == "published",
            Article.is_public.is_(True),
        )
    )
    article = result.scalar_one_or_none()
    if article is None:
        raise NotFoundException("文章不存在或非公开")
    article.view_count = (article.view_count or 0) + 1
    await db.flush()
    return _to_detail(article)


async def get_public_stats(db: AsyncSession) -> dict:
    public_filter = (Article.status == "published", Article.is_public.is_(True))
    total_articles = (await db.execute(select(func.count()).where(*public_filter))).scalar() or 0
    total_views = (
        await db.execute(select(func.coalesce(func.sum(Article.view_count), 0)).where(*public_filter))
    ).scalar() or 0
    category_rows = (
        await db.execute(
            select(Article.category, func.count(Article.id))
            .where(*public_filter, Article.category.is_not(None))
            .group_by(Article.category)
            .order_by(func.count(Article.id).desc())
        )
    ).all()
    recent = await db.execute(
        select(Article).where(*public_filter).order_by(Article.createdAt.desc()).limit(4)
    )
    categories = await db.execute(select(Category))
    category_ids = {category.name: category.id for category in categories.scalars().all()}
    popular_categories = [
        {"id": category_ids.get(name, 0), "name": name, "article_count": count}
        for name, count in category_rows[:5]
    ]
    return {
        "totalArticles": total_articles,
        "totalCategories": len(category_rows),
        "totalViews": total_views,
        "popularArticles": [_to_summary(article) for article in recent.scalars().all()],
        "popularCategories": popular_categories,
    }


async def get_article_by_id(
    db: AsyncSession, user_id: int, article_id: int, is_admin: bool = False
) -> ArticleDetail:
    """获取文章详情（同时增加浏览量）。admin 可查看任意用户的文章。"""
    filters = [Article.id == article_id]
    if not is_admin:
        filters.append(Article.user_id == user_id)
    result = await db.execute(select(Article).where(*filters))
    article = result.scalar_one_or_none()
    if article is None:
        raise NotFoundException("文章不存在")

    # 增加浏览量
    article.view_count = (article.view_count or 0) + 1
    await db.flush()
    return _to_detail(article)


async def get_blog_stats(db: AsyncSession, user_id: int, is_admin: bool = False) -> dict:
    """获取知识库统计信息。"""
    # 基础过滤条件
    article_filter = [Article.status == "published"]
    category_filter = []
    
    if not is_admin:
        article_filter.append(Article.user_id == user_id)
        category_filter.append(Category.user_id == user_id)

    # 1. 发布文章总数
    total_articles = (await db.execute(select(func.count()).where(*article_filter))).scalar() or 0

    # 2. 分类总数
    total_categories = (await db.execute(select(func.count()).where(*category_filter))).scalar() or 0

    # 3. 总浏览量
    total_views = (await db.execute(select(func.coalesce(func.sum(Article.view_count), 0)).where(*article_filter))).scalar() or 0

    # 4. 最近 4 篇文章
    recent_query = select(Article).where(*article_filter).order_by(Article.createdAt.desc()).limit(4)
    recent_result = await db.execute(recent_query)
    popular_articles = [_to_summary(a) for a in recent_result.scalars().all()]

    # 5. Top 5 分类
    cats_query = select(Category).where(*category_filter).order_by(Category.article_count.desc()).limit(5)
    cats_result = await db.execute(cats_query)
    popular_categories = [
        {"id": c.id, "name": c.name, "article_count": c.article_count}
        for c in cats_result.scalars().all()
    ]

    return {
        "totalArticles": total_articles,
        "totalCategories": total_categories,
        "totalViews": total_views,
        "popularArticles": popular_articles,
        "popularCategories": popular_categories,
    }


async def get_daily_stats(db: AsyncSession, user_id: int, is_admin: bool = False) -> list[dict]:
    """获取近 7 天每日文章创建统计。"""
    today = datetime.now().date()
    start_date = today - timedelta(days=6)

    query = select(
        func.date(Article.createdAt).label("date"),
        func.count().label("count"),
    ).where(
        func.date(Article.createdAt) >= start_date,
    )
    if not is_admin:
        query = query.where(Article.user_id == user_id)
    query = query.group_by(func.date(Article.createdAt))

    result = await db.execute(query)
    rows = result.all()

    # 构建 7 天数据（填充零）
    date_count: dict[str, int] = {}
    for row in rows:
        d = str(row.date) if hasattr(row, "date") else str(row[0])
        date_count[d] = row.count if hasattr(row, "count") else row[1]

    data = []
    for i in range(7):
        d = start_date + timedelta(days=i)
        ds = d.strftime("%Y-%m-%d")
        data.append({"date": ds, "count": date_count.get(ds, 0)})

    return data


async def create_article(db: AsyncSession, user_id: int, req: CreateArticleRequest) -> Article:
    """创建文章。"""
    now = datetime.now()
    article = Article(
        user_id=user_id,
        title=req.title,
        content=req.content,
        description=req.description,
        category=req.category or "随笔",
        tags=req.tags or [],
        cover_image=req.coverImage,
        view_count=0,
        status=req.status or "published",
        is_public=req.is_public if req.is_public is not None else False,
        createdAt=now,
        updatedAt=now,
    )
    db.add(article)
    await db.flush()

    # 更新分类计数
    await _update_category_count(db, article.category, user_id)

    # 自动建立向量索引
    if article.status == "published":
        try:
            from app.services import article_embedding_service
            await article_embedding_service.index_article(db, article)
        except Exception as e:
            logger.warning("[RAG] 自动索引新文章失败: %s", e)

    return article


async def update_article(
    db: AsyncSession,
    article_id: int,
    req: UpdateArticleRequest,
    user_id: int,
    is_admin: bool = False,
) -> Article:
    """更新文章。非 admin 仅能更新自己的文章。"""
    filters = [Article.id == article_id]
    if not is_admin:
        filters.append(Article.user_id == user_id)
    result = await db.execute(select(Article).where(*filters))
    article = result.scalar_one_or_none()
    if article is None:
        raise NotFoundException("文章不存在")

    old_category = article.category

    if req.title is not None:
        article.title = req.title
    if req.content is not None:
        article.content = req.content
    if req.description is not None:
        article.description = req.description
    if req.category is not None:
        article.category = req.category
    if req.tags is not None:
        article.tags = req.tags
    if req.coverImage is not None:
        article.cover_image = req.coverImage
    if req.status is not None:
        article.status = req.status
    if req.is_public is not None:
        article.is_public = req.is_public

    article.updatedAt = datetime.now()
    await db.flush()

    # 更新分类计数
    if req.category is not None and req.category != old_category:
        await _update_category_count(db, old_category, article.user_id)
        await _update_category_count(db, req.category, article.user_id)

    # 自动刷新向量索引
    from app.services import article_embedding_service
    if article.status == "published":
        try:
            await article_embedding_service.index_article(db, article)
        except Exception as e:
            logger.warning("[RAG] 自动刷新文章索引失败: %s", e)
    else:
        try:
            await article_embedding_service.remove_article(article_id, db)
        except Exception as e:
            logger.warning("[RAG] 自动移除未发布文章向量失败: %s", e)

    return article


async def delete_article(
    db: AsyncSession, article_id: int, user_id: int, is_admin: bool = False
) -> Article:
    """删除文章。非 admin 仅能删除自己的文章。"""
    filters = [Article.id == article_id]
    if not is_admin:
        filters.append(Article.user_id == user_id)
    result = await db.execute(select(Article).where(*filters))
    article = result.scalar_one_or_none()
    if article is None:
        raise NotFoundException("文章不存在")

    category = article.category
    user_id = article.user_id

    # 自动清理向量索引与切片
    try:
        from app.services import article_embedding_service
        await article_embedding_service.remove_article(article_id, db)
    except Exception as e:
        logger.warning("[RAG] 自动删除文章向量失败: %s", e)

    await db.delete(article)
    await db.flush()

    await _update_category_count(db, category, user_id)
    return article


async def _update_category_count(db: AsyncSession, category_name: str | None, user_id: int) -> None:
    """更新分类的文章计数。"""
    if not category_name:
        return

    count_result = await db.execute(
        select(func.count()).where(
            Article.category == category_name,
            Article.user_id == user_id,
            Article.status == "published",
        )
    )
    count = count_result.scalar() or 0

    cat_result = await db.execute(
        select(Category).where(Category.name == category_name, Category.user_id == user_id)
    )
    category = cat_result.scalar_one_or_none()

    if category:
        category.article_count = count
    else:
        now = datetime.now()
        new_cat = Category(
            user_id=user_id,
            name=category_name,
            article_count=count,
            createdAt=now,
            updatedAt=now,
        )
        db.add(new_cat)

    await db.flush()
